lib/spack/spack/build_environment.py

```diff
@@ -956,6 +956,8 @@ def add_modifications_for_dep(dep):
                 dpkg.setup_dependent_build_environment(env, spec)
             else:
                 dpkg.setup_dependent_run_environment(env, spec)
+            if dep in exe_deps:
```

Comment:
@scheibelp can you comment on this? I generally struggle to follow the logic in `modifications_from_dependencies`, but this does the job.

My issue though is that one line up we're calling `setup_dependent_run_environment` on all deps in `custom_mod_deps` which is `run_and_supporting_deps`, which includes *link* deps too... hence I added `if dep in exe_deps`. But generally this whole function is over complicated :/

Reply:
I think some link dependencies have internal run dependencies within the package that can require environment modifications -- I think we want to be doing the modifications for link deps too (although I guess we could argue about whether those should be listed as `link/run` deps).

Reply:
I do worry a little here about link deps whose behavior may be modified by environment variables (like libraries that use environment variables to find their config files).